### PR TITLE
fix(gatsby-plugin-mdx): support yarn PnP

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

See https://github.com/yarnpkg/berry/issues/689#issuecomment-574994256

In order to support yarn PnP, dependencies have to be explicit listed or required from their direct parents. `gatsby-plugin-mdx` has required some dependencies which are not listed in `package.json`, but sub-dependencies of `@mdx-js/mdx`.

We have 2 options to fix this:

1. List the dependencies explicitly. But we have to remember to update the range whenever `@mdx-js/mdx` updates. It would soon become troublesome and tedious.
2. Require the packages anyways, but require them from the `@mdx-js/mdx` package via `Module.createRequire`.

This PR goes for the second approach. The `createRequireFromPath` util is already implemented in the core package from `gatsby/src/utils/create-require-from-path.js`. We can just use it here. (Now that it's used by a plugin, should we move it to `gatsby-core-utils` or something similar?)

### Documentation

https://next.yarnpkg.com/features/pnp/, and https://next.yarnpkg.com/advanced/migration

## Related Issues

Related to https://github.com/yarnpkg/berry/issues/689. c.c. @arcanis

### CODE REVIEW

File: packages/gatsby-plugin-mdx/gatsby-node.js [hunk 1]
```diff
@@ -41,7 +41,7 @@ exports.preprocessSource = require(`./gatsby/preprocess-source`)
  */
 exports.onCreateBabelConfig = ({ actions }) => {
   actions.setBabelPlugin({
-    name: `@babel/plugin-proposal-object-rest-spread`,
+    name: require.resolve(`@babel/plugin-proposal-object-rest-spread`),
```
Comment (kevin940726) [inline packages/gatsby-plugin-mdx/gatsby-node.js]: Hmmm... maybe this is wrong, wonder if there’s a way to set the plugin this way. Might have to take a look at `createConfigItem`...

Comment (arcanis) [reply packages/gatsby-plugin-mdx/gatsby-node.js]: Babel plugins can/should be absolute paths, so this looks fine I think 🤔

Comment (kevin940726) [reply packages/gatsby-plugin-mdx/gatsby-node.js]: But now we cannot dedup it with other part of the app importing the same babel plugin, can we?
Maybe babel handle it nicely internally, but I’m not sure 🤔.

File: packages/gatsby-plugin-mdx/loaders/mdx-loader.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ const { getOptions } = require(`loader-utils`)
 const grayMatter = require(`gray-matter`)
 const unified = require(`unified`)
 const babel = require(`@babel/core`)
+const createRequireFromPath = require(`gatsby/dist/utils/create-require-from-path`)
```
Comment (pieh) [inline packages/gatsby-plugin-mdx/loaders/mdx-loader.js]: We should not import internals like that as they are not considered public API and can change at any time.

If there are utils from `gatsby` core package that other packages want to use, let's move that util to https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-core-utils and import this both in `gatsby` and in here

Comment (pieh) [reply packages/gatsby-plugin-mdx/loaders/mdx-loader.js]: (the move can be in the same PR)

Comment (kevin940726) [reply packages/gatsby-plugin-mdx/loaders/mdx-loader.js]: Sure, will do!

File: packages/gatsby-plugin-mdx/gatsby-node.js (hunk 1 above)
Comment (kevin940726) [reply packages/gatsby-plugin-mdx/gatsby-node.js]: Looks like there are already some same [usages](https://github.com/gatsbyjs/gatsby/blob/93f0645dc0088ef114eb6ca7ed05f23cae3556ce/packages/gatsby-plugin-typescript/src/gatsby-node.js#L4-L16) like this, I think it's fine!

**Comment by kevin940726** [comment]
yarn 2 has now been released! Any chance we can take a look at this again 😉? @pieh? (Sorry if I pinged the wrong person)

**Comment by pieh** [comment]
Just a note - I just published `gatsby-plugin-mdx@berry-mdx` with changes from this (+ bunch of other packages like `gatsby-core-utils`, `gatsby` etc) and will be testing this out. But you could already start using those dist-tags if you use Yarn 2 in meantime.

**Comment by pieh** [comment]
Ok, after few humps, I did get it to build it when using Yarn2 (see appendix :P) and I think we should get this in. @johno do you want to check changes to `gatsby-plugin-mdx` here? (I think they make total sense).


-- Yarn 2 appendix
- was getting:
```
YN0001: │ Error: fsevents@patch:fsevents@npm%3A2.0.7#builtin<compat/fsevents>::version=2.0.7&hash=e8cd9e: Cannot apply hunk #1
    at c (/Users/misiek/test/pr20638/.yarn/releases/yarn-berry.js:58:888149)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```
which I solved with `resolutions` by pinning `fsevents` to `2.1.2` (per https://github.com/yarnpkg/berry/issues/719)
- was getting:
```
Error: Something that got detected as your top-level application (because it doesn't seem to belong to any package) tried to access a package that is not declared in your dependencies
  Required package: minimatch (via "minimatch")
  Required by: /Users/misiek/test/pr20638/public/
  Require stack:
  - /Users/misiek/test/pr20638/public/render-page.js
  - /Users/misiek/test/pr20638/.yarn/unplugged/gatsby-virtual-1919d581a5/node_modules/gatsby/dist/utils/worker/render-html.js
  - /Users/misiek/test/pr20638/.yarn/unplugged/gatsby-virtual-1919d581a5/node_modules/gatsby/dist/utils/worker/child.js
  - /Users/misiek/test/pr20638/.yarn/cache/jest-worker-npm-24.9.0-bcd452e3a8-1.zip/node_modules/jest-worker/build/workers/processChild.js
```
which I hackily solved by installing `minimatch` in project (but that's something we will need to handle in core most likely, because this is where we generate `render-page.js` - tho I'm not sure if this is because of gatsby core or some other plugin that was in mdx starter that I was using as test case)

**Comment by pieh** [comment]
Published `gatsby-plugin-mdx@1.0.69`, thanks @kevin940726 for the fixes!

**Pull request merged**